File: client/lib/core/utils/subscription_calendar_days.dart
// IST calendar day counts for subscription windows (matches server
// subscriptionCalendarDays.js). Inclusive start/end; remaining counts
// from today (IST) through end, including today.

String _ymdIST(DateTime d) {
  final utc = d.toUtc();
  final ist = utc.add(const Duration(hours: 5, minutes: 30));
  return '${ist.year.toString().padLeft(4, '0')}-'
      '${ist.month.toString().padLeft(2, '0')}-'
      '${ist.day.toString().padLeft(2, '0')}';
}

String _istTodayYmd() => _ymdIST(DateTime.now());

/// Inclusive calendar days from subscription start → end (IST).
int totalDaysInclusiveIST(DateTime start, DateTime end) {
  final y1 = _ymdIST(start);
  final y2 = _ymdIST(end);
  final a = DateTime.parse('${y1}T00:00:00+05:30');
  final b = DateTime.parse('${y2}T00:00:00+05:30');
  final diff = b.difference(a).inDays;
  if (diff < 0) return 0;
  return diff + 1;
}

/// Days from today (IST) through end date (inclusive).
int remainingDaysInclusiveIST(DateTime start, DateTime end) {
  final startYmd = _ymdIST(start);
  final endYmd = _ymdIST(end);
  final todayYmd = _istTodayYmd();
  final remStart =
      startYmd.compareTo(todayYmd) >= 0 ? startYmd : todayYmd;
  if (remStart.compareTo(endYmd) > 0) return 0;
  final a = DateTime.parse('${remStart}T00:00:00+05:30');
  final b = DateTime.parse('${endYmd}T00:00:00+05:30');
  return b.difference(a).inDays + 1;
}

File: server/utils/subscriptionCalendarDays.js
/**
 * Subscription window day counts in Asia/Kolkata (IST), matching delivery rows.
 * Inclusive of both start and end calendar dates; "remaining" counts from
 * today (IST) through end, including today when the window is still active.
 */

const DAY_MS = 86400000;

/** IST calendar date string YYYY-MM-DD for "today". */
export function istTodayYmd() {
  return new Date().toLocaleDateString("en-CA", { timeZone: "Asia/Kolkata" });
}

/** YYYY-MM-DD in Asia/Kolkata for a stored Date. */
export function ymdIST(d) {
  if (!d) return "";
  return new Date(d).toLocaleDateString("en-CA", {
    timeZone: "Asia/Kolkata",
  });
}

/**
 * Inclusive calendar days from start → end (IST midnight boundaries).
 * Matches the delivery list row count for the subscription window.
 */
export function totalDaysInclusiveIST(startDate, endDate) {
  const startYmd = ymdIST(startDate);
  const endYmd = ymdIST(endDate);
  if (!startYmd || !endYmd) return 0;
  const startMs = new Date(`${startYmd}T00:00:00+05:30`).getTime();
  const endMs = new Date(`${endYmd}T00:00:00+05:30`).getTime();
  if (startMs > endMs) return 0;
  return Math.floor((endMs - startMs) / DAY_MS) + 1;
}

/**
 * Days left in the subscription window from today (IST) through end (inclusive).
 */
export function remainingDaysInclusiveIST(startDate, endDate) {
  const startYmd = ymdIST(startDate);
  const endYmd = ymdIST(endDate);
  if (!startYmd || !endYmd) return 0;
  const startMs = new Date(`${startYmd}T00:00:00+05:30`).getTime();
  const endMs = new Date(`${endYmd}T00:00:00+05:30`).getTime();
  const todayYmd = istTodayYmd();
  const todayMs = new Date(`${todayYmd}T00:00:00+05:30`).getTime();
  const remStartMs = Math.max(startMs, todayMs);
  let remainingDays = 0;
  if (remStartMs <= endMs) {
    for (let t = remStartMs; t <= endMs; t += DAY_MS) {
      remainingDays += 1;
    }
  }
  return remainingDays;
}
